File: scripts/rewrite-consumer-imports-to-semantic.py
import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
TARGET = ROOT / "restored"
MANIFEST = json.load(open(TARGET / ".deobfuscate-javascript/_full/manifest.json"))
IM = json.load(open(TARGET / "IMPORT_MAP.json"))

def get_class(b):
    info = MANIFEST["files"].get(b, {})
    return info.get("organization", {}).get("classification")

def target_exports(restored_path):
    p = TARGET / restored_path
    if p.is_dir():
        for name in ("index.ts", "index.tsx", "index.js", "index.jsx"):
            cand = p / name
            if cand.exists():
                p = cand
                break
    if not p.exists() or p.is_dir():
        return set()
    src = p.read_text()
    names = set()
    for m in re.finditer(
        r"\bexport\s+(?:declare\s+)?(?:const|function|class|type|interface)\s+([A-Za-z0-9_$]+)",
        src,
    ):
        names.add(m.group(1))
    for m in re.finditer(r"\bexport\s*\{([^}]+)\}", src):
        for part in m.group(1).split(","):
            part = part.strip()
            if not part:
                continue
            if " as " in part:
                names.add(part.split(" as ")[-1].strip())
            else:
                names.add(part)
    if re.search(r"\bexport\s+default\b", src):
        names.add("default")
    return names

IMPORT_RE = re.compile(
    r"import\s+\{([^}]+)\}\s+from\s+(['\"])([^'\"]+)\2"
)

fixed_files = 0
fixed_imports = 0
for cp in sorted((TARGET / ".deobfuscate-javascript/_full/checkpoints").glob("*")):
    src = cp.read_text()
    new_src = src
    for names_str, quote, source in IMPORT_RE.findall(src):
        if not source.startswith("./"):
            continue
        base = source[2:].split("/")[0]
        if "-" not in base:
            continue
        cls = get_class(base)
        if cls not in ("app-feature", "ui-component"):
            continue
        entry = IM.get("chunks", {}).get(base)
        if not entry or entry.get("status") != "done" or not entry.get("restored"):
            continue
        exports = entry.get("exports", {})
        if not exports:
            continue
        t_exports = target_exports(entry["restored"])
        # Build reverse map: lowercase original -> semantic
        rev = {k.lower(): v for k, v in exports.items()}
        replacements = {}
        for name in names_str.split(","):
            name = name.strip()
            if not name:
                continue
            # support "A as B"
            parts = [p.strip() for p in name.split(" as ")]
            imported = parts[0]
            alias = parts[1] if len(parts) > 1 else None
            if imported in t_exports:
                continue
            # try exact match in export map values
            semantic = None
            if imported in exports.values():
                semantic = imported
            else:
                # suffix match (case-insensitive) on original keys
                low = imported.lower()
                for orig, sem in exports.items():
                    if low.endswith(orig.lower()) and sem in t_exports:
                        semantic = sem
                        break
            if not semantic:
                continue
            if alias:
                replacements[name] = f"{semantic} as {alias}"
            else:
                replacements[name] = semantic
        if not replacements:
            continue
        # rebuild the names block preserving order/spacing roughly
        new_names = names_str
        for old, new in replacements.items():
            # replace whole specifier (handle comma spacing)
            pat = r"(^|,)\s*" + re.escape(old) + r"\s*(?=,|$)"
            new_names = re.sub(pat, lambda m: (m.group(1) or "") + new, new_names)
        # remove leading/trailing comma artifacts
        new_names = re.sub(r"^,\s*", "", new_names)
        new_names = re.sub(r"\s*,$", "", new_names)
        old_decl = f"import {{{names_str}}} from {quote}{source}{quote}"
        new_decl = f"import {{{new_names}}} from {quote}{source}{quote}"
        new_src = new_src.replace(old_decl, new_decl, 1)
        fixed_imports += len(replacements)
    if new_src != src:
        cp.write_text(new_src)
        fixed_files += 1

print(f"Rewrote {fixed_imports} imports in {fixed_files} checkpoint files.")

File: scripts/rewrite-consumer-imports-to-semantic.ts
import * as fs from "node:fs";
import * as path from "node:path";
import {
  rewriteSemanticImports,
  type SemanticImportMapping,
} from "../.agents/skills/deobfuscate-javascript/scripts/semantic-finalize.ts";

const ROOT = path.resolve(import.meta.dir, "..");
const TARGET = path.join(ROOT, "restored");
const MANIFEST = JSON.parse(
  fs.readFileSync(
    path.join(TARGET, ".deobfuscate-javascript/_full/manifest.json"),
    "utf-8",
  ),
);
const IM = JSON.parse(fs.readFileSync(path.join(TARGET, "IMPORT_MAP.json"), "utf-8"));

function getClass(b: string): string | undefined {
  return MANIFEST.files[b]?.organization?.classification;
}

function targetExports(restored: string): Set<string> {
  const p = path.join(TARGET, restored);
  if (!fs.existsSync(p)) return new Set();
  let file = p;
  if (fs.statSync(p).isDirectory()) {
    let found = false;
    for (const name of ["index.ts", "index.tsx", "index.js", "index.jsx"]) {
      const cand = path.join(p, name);
      if (fs.existsSync(cand)) {
        file = cand;
        found = true;
        break;
      }
    }
    if (!found) return new Set();
  }
  const src = fs.readFileSync(file, "utf-8");
  const names = new Set<string>();
  const declRe =
    /\bexport\s+(?:declare\s+)?(?:const|function|class|type|interface)\s+([A-Za-z0-9_$]+)/g;
  for (const m of src.matchAll(declRe)) names.add(m[1]);
  for (const m of src.matchAll(/\bexport\s*\{([^}]+)\}/g)) {
    for (const part of m[1].split(",")) {
      const trimmed = part.trim();
      if (!trimmed) continue;
      names.add(trimmed.includes(" as ") ? trimmed.split(" as ")[1].trim() : trimmed);
    }
  }
  if (/\bexport\s+default\b/.test(src)) names.add("default");
  return names;
}

const checkpointsDir = path.join(TARGET, ".deobfuscate-javascript/_full/checkpoints");
const files = fs.readdirSync(checkpointsDir).filter((f) => /\.[cm]?[jt]sx?$/.test(f));

let fixedFiles = 0;
let fixedImports = 0;

for (const file of files) {
  const cp = path.join(checkpointsDir, file);
  let source = fs.readFileSync(cp, "utf-8");
  const importRe = /import\s+\{([^}]+)\}\s+from\s+(['"])([^'"]+)\2/g;
  const mappings: SemanticImportMapping[] = [];
  for (const m of source.matchAll(importRe)) {
    const srcValue = m[3];
    if (!srcValue.startsWith("./")) continue;
    const base = srcValue.slice(2).split("/")[0];
    if (!base.includes("-")) continue;
    const cls = getClass(base);
    if (cls !== "app-feature" && cls !== "ui-component") continue;
    const entry = IM.chunks?.[base];
    if (!entry || entry.status !== "done" || !entry.restored) continue;
    const exports: Record<string, string> = entry.exports ?? {};
    if (Object.keys(exports).length === 0) continue;
    const tExports = targetExports(entry.restored as string);
    const renameMap: Record<string, string> = {};
    for (const spec of m[1].split(",")) {
      const trimmed = spec.trim();
      if (!trimmed) continue;
      const parts = trimmed.split(" as ").map((s) => s.trim());
      const imported = parts[0];
      if (imported === "default") continue;
      if (tExports.has(imported)) continue;
      let semantic: string | null = null;
      if (exports[imported]) semantic = exports[imported];
      else {
        const low = imported.toLowerCase();
        for (const [orig, sem] of Object.entries(exports)) {
          if (low.endsWith(orig.toLowerCase()) && tExports.has(sem)) {
            semantic = sem;
            break;
          }
        }
      }
      if (!semantic) continue;
      renameMap[imported] = semantic;
    }
    if (Object.keys(renameMap).length > 0) {
      mappings.push({ source: srcValue, exports: renameMap });
      fixedImports += Object.keys(renameMap).length;
    }
  }
  if (mappings.length === 0) continue;
  const rewritten = rewriteSemanticImports(source, mappings);
  if (rewritten !== source) {
    fs.writeFileSync(cp, rewritten);
    fixedFiles++;
  }
}

console.log(`Rewrote ${fixedImports} imports in ${fixedFiles} checkpoint files.`);
